library/ssl_srv.c

```diff
@@ -2793,26 +2793,35 @@ static int ssl_write_certificate_request( mbedtls_ssl_context *ssl )
      */
     if( ssl->minor_ver == MBEDTLS_SSL_MINOR_VERSION_3 )
     {
-        const int *cur;
-
         /*
          * Supported signature algorithms
          */
-        for( cur = ssl->conf->sig_hashes; *cur != MBEDTLS_MD_NONE; cur++ )
+        const uint16_t *sig_alg = mbedtls_ssl_get_sig_algs( ssl );
+        if( sig_alg == NULL )
+            return( MBEDTLS_ERR_SSL_BAD_CONFIG );
+
+        for( ; *sig_alg != MBEDTLS_TLS1_3_SIG_NONE; sig_alg++ )
         {
-            unsigned char hash = mbedtls_ssl_hash_from_md_alg( *cur );
+            /* High byte is hash */
+            unsigned char hash = MBEDTLS_BYTE_1( *sig_alg );
```

Comment:
I think we should move those checks into a `mbedtls_ssl_tls12_is_valid_sig_alg()` function (not necessarily this name though) that would be used in `mbedtls_ssl_write_sig_alg_ext` as well (see comment in this function).